### PR TITLE
Array marshalling support

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Implement support for marshalling blittable and non-blittable arrays, with the small size stack allocation optimization for arrays under 256 bytes.

Includes support for both SizeParamIndex and SizeConst with the same semantics as the runtime.

Example codegen:

Pass int array by value:

```csharp
namespace DllImportGenerator.IntegrationTests
{
    partial class NativeExportsNE
    {
        public partial class Arrays
        {
            public static partial int Sum(int[] values, int numValues)
            {
                unsafe
                {
                    int __retVal;
                    //
                    // Invoke
                    //
                    fixed (int *__values_gen_native = &System.Runtime.InteropServices.MemoryMarshal.GetArrayDataReference(values))
                    {
                        __retVal = Sum__PInvoke__(__values_gen_native, numValues);
                    }

                    return __retVal;
                }
            }

            [System.Runtime.InteropServices.DllImportAttribute("NativeExportsNE", EntryPoint = "sum_int_array")]
            extern private static unsafe int Sum__PInvoke__(int *values, int numValues);
        }
    }
}
```

Pass int array by reference:

```csharp
namespace DllImportGenerator.IntegrationTests
{
    partial class NativeExportsNE
    {
        public partial class Arrays
        {
            public static partial void Duplicate(ref int[] values, int numValues)
            {
                unsafe
                {
                    //
                    // Setup
                    //
                    int *__values_gen_native;
                    //
                    // Marshal
                    //
                    int values__bytelen = sizeof(int) * values.Length;
                    __values_gen_native = (int *)System.Runtime.InteropServices.Marshal.AllocCoTaskMem(values__bytelen);
                    values.AsSpan().CopyTo(new System.Span<int>(__values_gen_native, values.Length));
                    //
                    // Invoke
                    //
                    Duplicate__PInvoke__(&__values_gen_native, numValues);
                    //
                    // Unmarshal
                    //
                    if (__values_gen_native != null)
                    {
                        values = new int[checked((int)numValues)];
                        new System.Span<int>(__values_gen_native, values.Length).CopyTo(values);
                    }
                    else
                        values = null;
                    //
                    // Cleanup
                    //
                    System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__values_gen_native);
                }
            }

            [System.Runtime.InteropServices.DllImportAttribute("NativeExportsNE", EntryPoint = "duplicate_int_array")]
            extern private static unsafe void Duplicate__PInvoke__(int **values, int numValues);
        }
    }
}
```

Pass non-blittable array (string array) by reference:

```csharp
namespace DllImportGenerator.IntegrationTests
{
    partial class NativeExportsNE
    {
        public partial class Arrays
        {
            public static partial void ReverseStrings(ref string[] strArray, out int numElements)
            {
                unsafe
                {
                    numElements = default;
                    //
                    // Setup
                    //
                    ushort **__strArray_gen_native;
                    //
                    // Marshal
                    //
                    int strArray__bytelen = sizeof(ushort *) * strArray.Length;
                    __strArray_gen_native = (ushort **)System.Runtime.InteropServices.Marshal.AllocCoTaskMem(strArray__bytelen);
                    for (int i = 0; i < strArray.Length; ++i)
                    {
                        __strArray_gen_native[i] = (ushort *)System.Runtime.InteropServices.Marshal.StringToCoTaskMemUni(strArray[i]);
                    }

                    //
                    // Invoke
                    //
                    fixed (int *__numElements_gen_native = &numElements)
                    {
                        ReverseStrings__PInvoke__(&__strArray_gen_native, __numElements_gen_native);
                    }

                    //
                    // Unmarshal
                    //
                    if (__strArray_gen_native != null)
                    {
                        strArray = new string[checked((int)numElements)];
                        for (int i = 0; i < strArray.Length; ++i)
                        {
                            strArray[i] = __strArray_gen_native[i] == null ? null : new string ((char *)__strArray_gen_native[i]);
                        }
                    }
                    else
                        strArray = null;
                    //
                    // Cleanup
                    //
                    for (int i = 0; i < strArray.Length; ++i)
                    {
                        System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__strArray_gen_native[i]);
                    }

                    System.Runtime.InteropServices.Marshal.FreeCoTaskMem((System.IntPtr)__strArray_gen_native);
                }
            }

            [System.Runtime.InteropServices.DllImportAttribute("NativeExportsNE", EntryPoint = "reverse_strings")]
            extern private static unsafe void ReverseStrings__PInvoke__(ushort ***strArray, int *numElements);
        }
    }
}
```

Using SizeConst:

```csharp
namespace DllImportGenerator.IntegrationTests
{
    partial class NativeExportsNE
    {
        public partial class Arrays
        {
            public static partial byte[] GetLongBytes(long l)
            {
                unsafe
                {
                    byte[] __retVal;
                    //
                    // Setup
                    //
                    byte *__retVal_gen_native;
                    //
                    // Invoke
                    //
                    __retVal_gen_native = GetLongBytes__PInvoke__(l);
                    //
                    // Unmarshal
                    //
                    if (__retVal_gen_native != null)
                    {
                        __retVal = new byte[8];
                        new System.Span<byte>(__retVal_gen_native, __retVal.Length).CopyTo(__retVal);
                    }
                    else
                        __retVal = null;
                    return __retVal;
                }
            }

            [System.Runtime.InteropServices.DllImportAttribute("NativeExportsNE", EntryPoint = "get_long_bytes")]
            extern private static unsafe byte *GetLongBytes__PInvoke__(long l);
        }
    }
}
```

### CODE REVIEW

File: DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs [hunk 1]
```diff
@@ -48,7 +48,31 @@ CharEncoding CharEncoding
...
+    {
+        public MarshalAsInfo(UnmanagedType unmanagedType, CharEncoding charEncoding)
+            :this(unmanagedType, 0, 0, 0, charEncoding)
+        {
+        }
+
+        public MarshallingInfo CreateArraySubTypeMarshalAsInfo()
+        {
+            Debug.Assert(UnmanagedType is UnmanagedType.LPArray or UnmanagedType.ByValArray or UnmanagedType.SafeArray);
+            if (UnmanagedArraySubType == 0)
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs:61]: Boo. If we are going to use the type system with nullability, we also need to use it properly and accept that `0` isn't valid here. Which means we need a locally named constant with is used to indicate this state.

Comment (AaronRobinsonMSFT) [reply DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs:61]: The `UnspecifiedData` below is a good example.

File: DllImportGenerator/DllImportGenerator.IntegrationTests/ArrayTests.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+{
+    partial class NativeExportsNE
+    {
+        public partial class Arrays
+        {
+            [GeneratedDllImport(nameof(NativeExportsNE), EntryPoint = "sum_int_array")]
+            public static partial int Sum(int[] values, int numValues);
+            
+            [GeneratedDllImport(nameof(NativeExportsNE), EntryPoint = "duplicate_int_array")]
+            public static partial void Duplicate([MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 1)] ref int[] values, int numValues);
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator.IntegrationTests/ArrayTests.cs:19]: I don't see an `in` version.

File: DllImportGenerator/DllImportGenerator.IntegrationTests/ArrayTests.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            var expectedStrings = strings.Select(s => ReverseChars(s)).ToArray();
+            NativeExportsNE.Arrays.ReverseStrings(ref strings, out _);
+            
+            Assert.Equal((IEnumerable<string>)expectedStrings, strings);
+        }
+
+        [Fact]
+        public void ConstantSizeArray()
+        {
+            var longVal = 0x12345678L;
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator.IntegrationTests/ArrayTests.cs:104]: This is only defining 4 bytes, not 8. Is that intentional?

File: DllImportGenerator/DllImportGenerator.UnitTests/Compiles.cs [hunk 4]
```diff
@@ -33,21 +33,39 @@ public static IEnumerable<object[]> CodeSnippetsToCompile_NoDiagnostics()
...
+            yield return new[] { CodeSnippets.ArrayParametersAndModifiers<int>() };
+            yield return new[] { CodeSnippets.ArrayParametersAndModifiers<uint>() };
+            yield return new[] { CodeSnippets.ArrayParametersAndModifiers<long>() };
+            yield return new[] { CodeSnippets.ArrayParametersAndModifiers<ulong>() };
+            yield return new[] { CodeSnippets.ArrayParametersAndModifiers<float>() };
+            yield return new[] { CodeSnippets.ArrayParametersAndModifiers<double>() };
+            yield return new[] { CodeSnippets.ArrayParametersAndModifiers<bool>() };
+            yield return new[] { CodeSnippets.ArrayParametersAndModifiers<IntPtr>() };
+            yield return new[] { CodeSnippets.ArrayParametersAndModifiers<UIntPtr>() };
+            yield return new[] { CodeSnippets.ArrayParameterWithSizeParam<byte>(false) };
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator.UnitTests/Compiles.cs:49]: Let's use the argument name with free boolean constants here. Or perhaps we can leverage an existing enum to indicate the byref state?

File: DllImportGenerator/DllImportGenerator/DllImportGenerator.cs [hunk 5]
```diff
@@ -40,6 +40,9 @@ public void Execute(GeneratorExecutionContext context)
             var generatorDiagnostics = new GeneratorDiagnostics(context);
 
             var generatedDllImports = new StringBuilder();
+            generatedDllImports.AppendLine("#pragma warning disable CS8019");
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/DllImportGenerator.cs:43]: What's going on here?

File: DllImportGenerator/DllImportGenerator/Marshalling/BlittableArrayMarshaller.cs [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.Interop
+{
+    internal class BlittableArrayMarshaller : ConditionalStackallocMarshallingGenerator
+    {
+        private const int StackAllocBytesThreshold = 0x200;
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/BlittableArrayMarshaller.cs:12]: Document why we chose this number. If it was just a "reasonable guess" that is fine, but we need to know why.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallerHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.Interop
+{
+    internal static class MarshallerHelpers
+    {
+        public static ForStatementSyntax GetForLoop(string collectionIdentifier, string indexerIdentifier)
+        {
+            return ForStatement(EmptyStatement())
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallerHelpers.cs:14]: C# code example would be helpful here.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs [hunk 8]
```diff
@@ -263,5 +275,64 @@ private static IMarshallingGenerator CreateStringMarshaller(TypePositionInfo inf
 
             throw new MarshallingNotSupportedException(info, context);
         }
+
+
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:279]: Extra line.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs [hunk 9]
```diff
@@ -156,23 +158,33 @@ internal class MarshallingGenerators
...
+                case { MarshallingAttributeInfo: BlittableTypeAttributeInfo }:
                     return Blittable;
 
                 // Marshalling in new model
                 case { MarshallingAttributeInfo: NativeMarshallingAttributeInfo marshalInfo }:
                     return Forwarder;
 
                 // Simple marshalling with new attribute model, only have type name.
-                case { MarshallingAttributeInfo: GeneratedNativeMarshallingAttributeInfo { NativeMarshallingFullyQualifiedTypeName: string name } }:
+                case { MarshallingAttributeInfo: GeneratedNativeMarshallingAttributeInfo(string nativeTypeName) }:
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:172]: I don't understand the `string nativeTypeName` here. Why is that there?

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs [hunk 10]
```diff
@@ -263,5 +275,64 @@ private static IMarshallingGenerator CreateStringMarshaller(TypePositionInfo inf
 
             throw new MarshallingNotSupportedException(info, context);
         }
+
+        
+        private static ExpressionSyntax GetNumElementsExpressionFromMarshallingInfo(TypePositionInfo info, StubCodeContext context)
+        {
+            ExpressionSyntax numElementsExpression;
+            if (info.MarshallingAttributeInfo is MarshalAsInfo { ArraySizeParamIndex: short index, ArraySizeConst: int constSize })
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:283]: I really dislike this deeply nested local declarations. They aren't clearer to me but rather seem to be creating side-effects. Let's just make them local when we end the if branch.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs [hunk 11]
```diff
@@ -263,5 +275,64 @@ private static IMarshallingGenerator CreateStringMarshaller(TypePositionInfo inf
...
+                LiteralExpressionSyntax? constSizeExpression = constSize != MarshalAsInfo.UnspecifiedData
+                    ? LiteralExpression(SyntaxKind.NumericLiteralExpression, Literal(constSize))
+                    : null;
+                ExpressionSyntax? sizeParamIndexExpression = null;
+                if (index != MarshalAsInfo.UnspecifiedData)
+                {
+                    TypePositionInfo? paramIndexInfo = context.GetTypePositionInfoForManagedIndex(index);
+                    if (paramIndexInfo is null)
+                    {
+                        throw new MarshallingNotSupportedException(info, context);
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:294]: Can we make this exception more precise here. I think this really means the caller didn't provide an index correct? If this will be a specific message then okay, otherwise we should provide details.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs [hunk 12]
```diff
@@ -263,5 +275,64 @@ private static IMarshallingGenerator CreateStringMarshaller(TypePositionInfo inf
...
+                    }
+                    else if (!paramIndexInfo.ManagedType.IsIntegralType())
+                    {
+                        throw new MarshallingNotSupportedException(info, context);
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:298]: Same as above. Is this a generic message?

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs [hunk 13]
```diff
@@ -263,5 +275,64 @@ private static IMarshallingGenerator CreateStringMarshaller(TypePositionInfo inf
...
+                        var (managed, native) = context.GetIdentifiers(paramIndexInfo);
+                        string identifier = Create(paramIndexInfo, context).UsesNativeIdentifier(paramIndexInfo, context) ? native : managed;
+                        sizeParamIndexExpression = CastExpression(
+                                PredefinedType(Token(SyntaxKind.IntKeyword)),
+                                IdentifierName(identifier));
+                    }
+                }
+                numElementsExpression = (constSizeExpression, sizeParamIndexExpression) switch
+                {
+                    (null, null) => throw new MarshallingNotSupportedException(info, context),
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:311]: Specific message.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs [hunk 14]
```diff
@@ -263,5 +275,64 @@ private static IMarshallingGenerator CreateStringMarshaller(TypePositionInfo inf
...
+        {
+            var elementMarshaller = Create(TypePositionInfo.CreateForType(elementType, elementMarshallingInfo), context);
+            ExpressionSyntax numElementsExpression = LiteralExpression(SyntaxKind.NumericLiteralExpression, Literal(0));
+            if (info.IsManagedReturnPosition || (info.IsByRef && info.RefKind != RefKind.In))
+            {
+                // In this case, we need a numElementsExpression supplied from metadata, so we'll calculate it here.
+                numElementsExpression = GetNumElementsExpressionFromMarshallingInfo(info, context);
+            }
+            
+            return elementMarshaller == Blittable ? new BlittableArrayMarshaller(numElementsExpression) : new NonBlittableArrayMarshaller(elementMarshaller, numElementsExpression);
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:335]: Just for style can we put this on multiple lines?

File: DllImportGenerator/DllImportGenerator/Marshalling/NonBlittableArrayMarshaller.cs [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.Interop
+{
+    internal class NonBlittableArrayMarshaller : ConditionalStackallocMarshallingGenerator
+    {
+        private const int StackAllocBytesThreshold = 0x200;
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/Marshalling/NonBlittableArrayMarshaller.cs:12]: Constants should have documentation for "why"

File: DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs [hunk 16]
```diff
@@ -48,7 +48,31 @@ CharEncoding CharEncoding
...
+                return NoMarshallingInfo.Instance;
+            }
+            return new MarshalAsInfo(UnmanagedArraySubType, CharEncoding);
+        }
+
+        /// <summary>
+        /// Helper method to enable cleaner pattern matching for the common case of
+        /// a MarshalAs attribute that just uses the constructor parameter and no additional properties.
+        /// </summary>
+        public void Deconstruct(out UnmanagedType unmanagedType) => unmanagedType = UnmanagedType;
```
Comment (AaronRobinsonMSFT) [inline DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs:72]: This doesn't seem to be used anywhere.

File: DllImportGenerator/DllImportGenerator/DllImportGenerator.cs (hunk 5 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/DllImportGenerator.cs:43]: I added a default `using System;` to enable using the `.AsSpan()` extension method on arrays in the blittable array marshalling. This pragma disables the "unused using" warning when the namespace isn't used.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs (hunk 9 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:172]: This statement has the same meaning as the previous version, but instead of using the property deconstruction syntax, it uses the positional syntax that matches the record's constructor (well, technically the auto-generated `Deconstruct` method that matches the primary constructor). So this case matches a `GeneratedNativeMarshallingAttributeInfo` attribute that was created with any `string`. That `string` is then stored in the `nativeTypeName` variable.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs (hunk 10 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:283]: I personally think this way is cleaner, but I can rearrange things and gather these properties in a more traditional C# style.

File: DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs (hunk 16 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs:72]: The `Deconstruct` method is  a special method name for pattern matching. In this case, it enables the `x is MarshalAsInfo(UnmanagedType.I4)` syntax instead of the `x is MarshalAsInfo { UnmanagedType: UnmanagedType.I4 }`. The constructor style syntax is much cleaner in my opinion.

File: DllImportGenerator/DllImportGenerator/StubCodeContext.cs [hunk 17]
```diff
@@ -66,6 +66,8 @@ public enum Stage
 
         public abstract bool StackSpaceUsable { get; }
 
+        public abstract bool CanUseAdditionalTemporaryState { get; }
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/StubCodeContext.cs:69]: This bool encapsulates an implicit policy that we had before and are using with the conditional stackalloc and the SafeHandle marshalling. When this is false, we can't introduce additional variables to track state; we can only use the `managedIdentifier` and `nativeIdentifier` variables to store state.

File: DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs (hunk 16 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs:72]: Now that we have the two-parameter constructor with the CharSet however, I might modify this method to instead mirror the second constructor with the CharSet parameter.

File: DllImportGenerator/DllImportGenerator/StubCodeContext.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (elinor-fung) [reply DllImportGenerator/DllImportGenerator/StubCodeContext.cs:69]: Can we list out the matrix of scenarios / expected values for these properties - in `Pipeline.md` and/or as a comment here? Specifically `PinningSupported`, `StackSpaceUsable`, and `CanUseAdditionalTemporaryState`.

[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/StubCodeContext.cs:69]: I'll update Pipeline.md to include a description of these switches and what they describe as supported.

File: DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs (hunk 16 above)
Comment (AaronRobinsonMSFT) [reply DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs:72]: When/how often is this called? A very common complaint about C++ is implicit constructors and this seems very similar in context. I personally am not fan of the current switch statement we have because it is no longer a jump table but rather a much more complex comparison matrix. @elinor-fung removed the `char` and `string` because it was unnecessarily complex and this seems to add to enable that pattern which is far too dense in my experience What are we gaining with this pattern? I am close to saying this isn't an approach we should be using given it is much clearer to have specialized functions that are easier to maintain than the enormous switch statement.

Is there an argument for this approach?

File: DllImportGenerator/DllImportGenerator/DllImportGenerator.cs (hunk 5 above)
Comment (AaronRobinsonMSFT) [reply DllImportGenerator/DllImportGenerator/DllImportGenerator.cs:43]: Let's comment that. Thank you making the code clean. We need to make sure we document why we are disabling anything.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs (hunk 9 above)
Comment (AaronRobinsonMSFT) [reply DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:172]: I think I understood that actually - yay for being in on the new C# stuff. What I don't see is where `nativeTypeName` is being used, that is my main confusion.

File: DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs (hunk 16 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs:72]: The main argument for adding the `Deconstruct` method is to make all of the patterns that were of the form `MarshalAsInfo { UnmanagedType: UnmanagedType.I4 }` to instead be of the form `MarshalAsInfo(UnmanagedType.I4)`. In our marshaller selector in this PR, it's being used 15 times. This syntax also enables the simple MarshalAs checks that just depend on the first parameter to the MarshalAs attribute to look like the application of the MarshalAs attribute, which is a nice touch.

If I move some of the properties (such as the array-specific ones) on the MarshalAs record from being positional (in the constructor) to just regular init properties, I can remove the need for this Deconstruct method by making the two-parameter version be auto-generated. I'll try a few things out to make the Deconstruct method unnecessary.

Comment (AaronRobinsonMSFT) [reply DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs:72]: Appreciate the effort.

File: DllImportGenerator/DllImportGenerator/DllImportGenerator.cs (hunk 5 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/DllImportGenerator.cs:43]: We could just cast to manually create the span, but I like the `AsSpan` method because it's cleaner and because emitting a generic type in Roslyn is even more verbose than usual. I can change it to instead just create the span instance with the constructor.

File: DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs (hunk 9 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:172]: It's not being used today. I originally added `nativeTypeName` when I was starting work on the non-blittable struct marshaller. It will be used when we support generating code for non-blittable structs to enable us to have both source generators working without a dependency between the two (which isn't currently allowed). `nativeTypeName` will be a pre-calculated name for what the struct marshalling source generator will name the native type.

Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:172]: I can remove it and just match on `GeneratedNativeMarshallingAttribute` until we implement support if you want.

Comment (AaronRobinsonMSFT) [reply DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs:172]: Not necessary. Wanted to make sure we knew why.

**Comment by jkoritzinsky** [comment]
Any more feedback on this PR?

File: DllImportGenerator/TestAssets/NativeExports/Arrays.cs [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+        [UnmanagedCallersOnly(EntryPoint = "reverse_strings")]
+        public static void ReverseStrings(ushort*** strArray, int* numValues)
+        {
+            List<IntPtr> newStrings = new List<IntPtr>();
+            for (int i = 0; (nint)(*strArray)[i] != 0; i++)
+            {
+                newStrings.Add((IntPtr)Strings.Reverse((*strArray)[i]));
+            }
+            newStrings.Add(IntPtr.Zero);
+            *strArray = (ushort**)Marshal.AllocCoTaskMem(sizeof(ushort*) * newStrings.Count);
```
Comment (elinor-fung) [inline DllImportGenerator/TestAssets/NativeExports/Arrays.cs:90]: Do we need to free the existing array first?

File: DllImportGenerator/DllImportGenerator/Marshalling/ConditionalStackallocMarshallingGenerator.cs [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+
+        private static string GetByteLengthIdentifier(string managedIdentifier) => $"{managedIdentifier}__bytelen";
+
+        private static string GetStackAllocIdentifier(string managedIdentifier) => $"{managedIdentifier}__stackptr";
+
+        protected IEnumerable<StatementSyntax> GenerateConditionalAllocationSyntax(
+            TypePositionInfo info, 
+            StubCodeContext context,
+            int stackallocMaxSize,
+            bool checkForNull)
```
Comment (elinor-fung) [inline DllImportGenerator/DllImportGenerator/Marshalling/ConditionalStackallocMarshallingGenerator.cs:21]: All current usages seem to set this to true - when would we not want the null check?

File: DllImportGenerator/DllImportGenerator/Marshalling/NonBlittableArrayMarshaller.cs [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+                        foreach (var statement in GenerateConditionalAllocationSyntax(
+                            info,
+                            context,
+                            StackAllocBytesThreshold,
+                            checkForNull: true))
+                        {
+                            yield return statement;
+                        }
+
+                        string indexerIdentifier = "i";
```
Comment (elinor-fung) [inline DllImportGenerator/DllImportGenerator/Marshalling/NonBlittableArrayMarshaller.cs:86]: Quick comment that this is iterating over and marshalling each element?

File: DllImportGenerator/DllImportGenerator/Marshalling/NonBlittableArrayMarshaller.cs [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+                        var arraySubContext = new ArrayMarshallingCodeContext(context.CurrentStage, indexerIdentifier, context);
+                        yield return MarshallerHelpers.GetForLoop(managedIdentifer, indexerIdentifier)
```
Comment (elinor-fung) [inline DllImportGenerator/DllImportGenerator/Marshalling/NonBlittableArrayMarshaller.cs:88]: Null check before doing the loop too, since it does `<identifier>.Length`? Also in Cleanup.

Comment (elinor-fung) [reply DllImportGenerator/DllImportGenerator/Marshalling/NonBlittableArrayMarshaller.cs:88]: We should also have tests with null array parameter in / out.

File: DllImportGenerator/DllImportGenerator/Marshalling/BlittableArrayMarshaller.cs [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+                SizeOfExpression(GetElementTypeSyntax(info)),
+                MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression,
+                    IdentifierName(context.GetIdentifiers(info).managed),
+                    IdentifierName("Length")
+                ));
+        }
+
+        protected override StatementSyntax GenerateStackallocOnlyValueMarshalling(TypePositionInfo info, StubCodeContext context, SyntaxToken byteLengthIdentifier, SyntaxToken stackAllocPtrIdentifier)
+        {
+            return EmptyStatement();
```
Comment (elinor-fung) [inline DllImportGenerator/DllImportGenerator/Marshalling/BlittableArrayMarshaller.cs:219]: Will this actually show up as an empty statement (`;`) in the generated code?

**Review by elinor-fung — APPROVED** [review]

File: DllImportGenerator/DllImportGenerator/Marshalling/BlittableArrayMarshaller.cs (hunk 22 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/Marshalling/BlittableArrayMarshaller.cs:219]: Yes it will.
